File: packages/sdk-py/examples/deploy_and_wait.py
"""Build + deploy + wait for RUNNING state.

Usage::

    python examples/deploy_and_wait.py <service_id> <git_sha> [environment]

Environment variables:
    ENCLII_TOKEN     Required. API token.
    ENCLII_API_URL   Optional. Defaults to https://api.enclii.dev.
"""

from __future__ import annotations

import asyncio
import os
import sys

from enclii_sdk import AsyncEncliiClient


async def main(service_id: str, git_sha: str, environment: str = "production") -> int:
    async with AsyncEncliiClient(
        base_url=os.environ.get("ENCLII_API_URL", "https://api.enclii.dev"),
        token=os.environ["ENCLII_TOKEN"],
    ) as enclii:
        print(f"Triggering build for {service_id} @ {git_sha}…")
        release = await enclii.services.build(service_id, git_sha=git_sha)
        print(f"  release_id={release.id} status={release.status}")

        # Poll the release until it's READY or FAILED. The deploy below
        # will 400 if we try to use a not-yet-built release.
        while release.status.value == "building":
            await asyncio.sleep(5)
            releases = await enclii.services.list_releases(service_id)
            for r in releases:
                if r.id == release.id:
                    release = r
                    break
            print(f"  build status: {release.status}")

        if release.status.value != "ready":
            print(f"Build failed: {release.error_message}")
            return 1

        print(f"Deploying {release.id} to {environment}…")
        deployment = await enclii.services.deploy(
            service_id,
            release_id=str(release.id),
            environment_name=environment,
        )
        print(f"  deployment_id={deployment.id}")

        print("Waiting for RUNNING state…")
        final = await enclii.deployments.wait_for_running(
            str(deployment.id), timeout=600.0
        )
        print(f"Deployed {final.version_label} ({final.ready_replicas} replicas)")
        return 0


if __name__ == "__main__":
    if len(sys.argv) < 3:
        print(__doc__)
        sys.exit(2)
    sys.exit(
        asyncio.run(
            main(sys.argv[1], sys.argv[2], sys.argv[3] if len(sys.argv) > 3 else "production")
        )
    )

File: packages/sdk-ts/examples/deploy-and-wait.ts
/**
 * Deploy a release and wait until the pods are running.
 *
 * Usage:
 *   ENCLII_TOKEN=... ENCLII_SERVICE_ID=... ENCLII_RELEASE_ID=... \
 *     tsx examples/deploy-and-wait.ts
 */

import { EncliiClient, NotFoundError } from '@madfam/enclii-sdk';

async function main() {
  const client = new EncliiClient({
    baseUrl: process.env.ENCLII_BASE_URL ?? 'https://api.enclii.dev/v1',
    token: process.env.ENCLII_TOKEN,
  });

  const serviceId = required('ENCLII_SERVICE_ID');
  const releaseId = required('ENCLII_RELEASE_ID');
  const environment = process.env.ENCLII_ENV ?? 'dev';

  console.log(`Deploying release ${releaseId} to ${serviceId} in ${environment}...`);
  const dep = await client.deployments.deploy(serviceId, {
    release_id: releaseId,
    environment_name: environment,
  });
  console.log(`Deployment ${dep.id} created (v${dep.version_number ?? '?'})`);

  try {
    const final = await client.deployments.wait(dep.id, {
      intervalMs: 5_000,
      timeoutMs: 10 * 60_000,
    });
    if (final.status === 'running') {
      console.log(`Success. Health=${final.health}`);
      process.exit(0);
    }
    console.error(`Deployment ended in ${final.status}: ${final.error_message}`);
    process.exit(1);
  } catch (err) {
    if (err instanceof NotFoundError) {
      console.error(`Deployment not found — may have been deleted: ${err.message}`);
    } else {
      console.error(`Deploy failed: ${(err as Error).message}`);
    }
    process.exit(2);
  }
}

function required(name: string): string {
  const v = process.env[name];
  if (!v) {
    console.error(`missing required env var: ${name}`);
    process.exit(64);
  }
  return v;
}

main();
